<img style="float: left; margin: 30px 15px 15px 15px;" src="https://pngimage.net/wp-content/uploads/2018/06/logo-iteso-png-5.png" width="300" height="500" /> 
    
    
### <font color='navy'> Simulación de procesos financieros. 

**Nombres:** Flavio Cesar Palacios Salas y Moisés Flores Ortiz
              

**Fecha:** 17 de febrero del 2021.

**Expediente** : Moisés Flores Ortiz: 722183. Flavio Cesar Palacios Salas: 729825.
**Profesor:** Oscar David Jaramillo Zuluaga.

# Tarea 3: Clase 8

### Link de Github
[Github](https://github.com/Palacios-F/ProyectoConjunto_FCPalacios_MFlores)

## Instrucciones
> Usando compresión de listas o funciones map(sino recuerda como funciona observar el siguiente enlace https://www.pythonforbeginners.com/lists/list-comprehensions-in-python/), resolver los siguientes ejercicios:

## Ejercicios
>1. Resolver la siguiente ecuación recursiva usando funciones como se vió en clase
$$ D_{n}=(n-1) D_{n-1}+(n-1) D_{n-2} \quad n\ge 3$$
con $D_1=0$ y $D_2 = 1$

### Solución 1 Flavio Palacios

In [1]:
import numpy as np
def recur(d1:'number in the first position',d2:'number in the second position',n:'number to calculate'):
    """"
    Function to calculate the dn term given of the next recursive fucntion given
    the two initial numbers
    D_{n}=(n-1) D_{n-1}+(n-1) D_{n-2}  for n>= 3
    """
    d = np.zeros(n)
    d[0] = d1
    d[1] = d2 
    def vector(i):
        nonlocal d
        d[i-1] = (i-1)*(d[i-2]+d[i-3])
    [vector(i) for i in range(3,n+1)]
    return d[-1]
recur(0,1,5)

44.0

### Solución 1 Moisés Flores

In [1]:
##### Ejercicio 1
import numpy as np
def recursive_method1(n:"Número de elementos"):
    '''
    Esta es la programación de la ecuación recursiva  D_{n}=(n-1) D_{n-1}+(n-1) D_{n-2} \quad n\ge 3
    con D_1=0 y D_2 = 1
    ----------
    N:'cantidad de términos'
    '''
    D=np.zeros(n)
    D[0] = 0; D[1] = 1;
    def llenar_vector(i):
        nonlocal D
        D[i-1] = (i-1)*D[i-2]+(i-1)*D[i-3]
    [llenar_vector(i) for i in range(3,n+1)]
    return np.array(D)
recursive_method1(6)

array([  0.,   1.,   2.,   9.,  44., 265.])

### Respuesta 
 La respuesta es:
 
 <font color="red">$[  0.,   1.,   2.,   9.,  44., 265.]$</font>

>2. Count the number of spaces in the following string `variable =  relaciónn  requiere,  para obtener  el  valor  de un  cierto  término,  el conocimiento de los dos anteriores`.

### Solución 2 Flavio Palacios

In [3]:
string = 'variable =  relaciónn  requiere,  para obtener  el  valor  de un  cierto  término,  el conocimiento de los dos anteriores'
from functools import reduce
len(string) - reduce(lambda x,y: x+y,[len(i)for i in string.split()])

26

### Solución 2 Moisés Flores

In [2]:
###### Ejercicio 2
variable ="  relaciónn  requiere,  para obtener  el  valor  de un  cierto  término,  el conocimiento de los dos anteriores"
def ejercicio_2(string):
    return len(list(filter(lambda x: x==" ",variable)))
ejercicio_2(variable)

25

### Respuesta 
 La respuesta es:
 
 <font color="red">$25$</font>

>3. Remove all of the vowels in a string [make a list of the non-vowels].

### Solución 3 Flavio Palacios

In [6]:
z = ['a','e','i','o','u']
a3 = [list(filter(lambda x:x != '',[s[i] if s[i] not in z else '' for i in range(len(s))])) for s in string.split()]
a3

[['v', 'r', 'b', 'l'],
 ['='],
 ['r', 'l', 'c', 'ó', 'n', 'n'],
 ['r', 'q', 'r', ','],
 ['p', 'r'],
 ['b', 't', 'n', 'r'],
 ['l'],
 ['v', 'l', 'r'],
 ['d'],
 ['n'],
 ['c', 'r', 't'],
 ['t', 'é', 'r', 'm', 'n', ','],
 ['l'],
 ['c', 'n', 'c', 'm', 'n', 't'],
 ['d'],
 ['l', 's'],
 ['d', 's'],
 ['n', 't', 'r', 'r', 's']]

### Solución 3 Moisés Flores

In [3]:
###### Ejercicio 3
import string as stg
variable ="  relaciónn  requiere,  para obtener  el  valor  de un  cierto  término,  el conocimiento de los dos anteriores"
def ejercicio_3(string):
    ###### Primero se quitan los espacios en blanco
    string=''.join(string.split())
    ###### Después se quitan todos los digítos posibles
    string=''.join(ch for ch in string if ch not in stg.punctuation)
    #### Se hace una lista con vocales con y sin tilde
    vowels_list = ['a','e','i','o','u',
                   'á','é','í','ó','ú']
    #### Se hace una función para filtrar las palabras que no están es la lista vowels_list
    return list(filter(lambda x:x if not x in vowels_list else False,string))
ejercicio_3(variable)

['r',
 'l',
 'c',
 'n',
 'n',
 'r',
 'q',
 'r',
 'p',
 'r',
 'b',
 't',
 'n',
 'r',
 'l',
 'v',
 'l',
 'r',
 'd',
 'n',
 'c',
 'r',
 't',
 't',
 'r',
 'm',
 'n',
 'l',
 'c',
 'n',
 'c',
 'm',
 'n',
 't',
 'd',
 'l',
 's',
 'd',
 's',
 'n',
 't',
 'r',
 'r',
 's']

### Respuesta 
 La respuesta es:
 
 <font color="red">$['r',
 'l',
 'c',
 'n',
 'n',
 'r',
 'q',
 'r',
 'p',
 'r',
 'b',
 't',
 'n',
 'r',
 'l',
 'v',
 'l',
 'r',
 'd',
 'n',
 'c',
 'r',
 't',
 't',
 'r',
 'm',
 'n',
 'l',
 'c',
 'n',
 'c',
 'm',
 'n',
 't',
 'd',
 'l',
 's',
 'd',
 's',
 'n',
 't',
 'r',
 'r',
 's']$</font>

>4. Find all of the words in a string that are less than 4 letters.

### Solución 4 Flavio Palacios

In [5]:
a4 = list(filter(lambda st: st if len(st) <= 4 else "",string.split()))
a4

['=', 'para', 'el', 'de', 'un', 'el', 'de', 'los', 'dos']

### Solución 4 Moisés Flores

In [4]:
###### Ejercicio 4
variable ="  relaciónn  requiere,  para obtener  el  valor  de un  cierto  término,  el conocimiento de los dos anteriores"
def ejercicio_4(string):
    return list(filter(lambda x: len(x) < 4,variable.split()))
ejercicio_4(variable)

['el', 'de', 'un', 'el', 'de', 'los', 'dos']

### Respuesta 
 La respuesta es:
 
 <font color="red">$['el', 'de', 'un', 'el', 'de', 'los', 'dos']$</font>

>5. Use a dictionary comprehension to count the length of each word in a sentence.

### Solución 5 Flavio Palacios

In [7]:
l = 'Ejemplo de una frase'
lw = {word:len(word) for i, word in enumerate(l.split())}
lw

{'Ejemplo': 7, 'de': 2, 'una': 3, 'frase': 5}

### Solución 5 Moisés Flores

In [5]:
####### Ejercicio 5
variable ="  relaciónn  requiere,  para obtener  el  valor  de un  cierto  término,  el conocimiento de los dos anteriores"
def ejercicio_5(string):
     return {word: len(word) for word in string.split()}
ejercicio_5(variable)

{'relaciónn': 9,
 'requiere,': 9,
 'para': 4,
 'obtener': 7,
 'el': 2,
 'valor': 5,
 'de': 2,
 'un': 2,
 'cierto': 6,
 'término,': 8,
 'conocimiento': 12,
 'los': 3,
 'dos': 3,
 'anteriores': 10}

### Respuesta 
 La respuesta es:
 
 <font color="red">$ {'relaciónn': 9,
 'requiere,': 9,
 'para': 4,
 'obtener': 7,
 'el': 2,
 'valor': 5,
 'de': 2,
 'un': 2,
 'cierto': 6,
 'término,': 8,
 'conocimiento': 12,
 'los': 3,
 'dos': 3,
 'anteriores': 10} $</font>

>6. Use a nested list comprehension to find all of the numbers from 1-1000 that are divisible by any single digit besides 1 (2-9). 

### Solución 6 Flavio Palacios

In [10]:
num = np.arange(2,10)
arr = np.arange(1,1001)
comp = lambda x,i: x if x % i == 0 else None
v = [np.array(list(filter(lambda x: x is not None,([comp(element,number) for element in arr])))) for number in num]
v

[array([   2,    4,    6,    8,   10,   12,   14,   16,   18,   20,   22,
          24,   26,   28,   30,   32,   34,   36,   38,   40,   42,   44,
          46,   48,   50,   52,   54,   56,   58,   60,   62,   64,   66,
          68,   70,   72,   74,   76,   78,   80,   82,   84,   86,   88,
          90,   92,   94,   96,   98,  100,  102,  104,  106,  108,  110,
         112,  114,  116,  118,  120,  122,  124,  126,  128,  130,  132,
         134,  136,  138,  140,  142,  144,  146,  148,  150,  152,  154,
         156,  158,  160,  162,  164,  166,  168,  170,  172,  174,  176,
         178,  180,  182,  184,  186,  188,  190,  192,  194,  196,  198,
         200,  202,  204,  206,  208,  210,  212,  214,  216,  218,  220,
         222,  224,  226,  228,  230,  232,  234,  236,  238,  240,  242,
         244,  246,  248,  250,  252,  254,  256,  258,  260,  262,  264,
         266,  268,  270,  272,  274,  276,  278,  280,  282,  284,  286,
         288,  290,  292,  294,  296, 

### Solución 6 Moisés Flores

In [6]:
###### Ejercicio_6
import numpy as np
def ejercicio_6():
    return np.array([i for i in range(1,1001) if [j for j in range(2,10) if i%j == 0]])
ejercicio_6()

array([   2,    3,    4,    5,    6,    7,    8,    9,   10,   12,   14,
         15,   16,   18,   20,   21,   22,   24,   25,   26,   27,   28,
         30,   32,   33,   34,   35,   36,   38,   39,   40,   42,   44,
         45,   46,   48,   49,   50,   51,   52,   54,   55,   56,   57,
         58,   60,   62,   63,   64,   65,   66,   68,   69,   70,   72,
         74,   75,   76,   77,   78,   80,   81,   82,   84,   85,   86,
         87,   88,   90,   91,   92,   93,   94,   95,   96,   98,   99,
        100,  102,  104,  105,  106,  108,  110,  111,  112,  114,  115,
        116,  117,  118,  119,  120,  122,  123,  124,  125,  126,  128,
        129,  130,  132,  133,  134,  135,  136,  138,  140,  141,  142,
        144,  145,  146,  147,  148,  150,  152,  153,  154,  155,  156,
        158,  159,  160,  161,  162,  164,  165,  166,  168,  170,  171,
        172,  174,  175,  176,  177,  178,  180,  182,  183,  184,  185,
        186,  188,  189,  190,  192,  194,  195,  1

### Respuesta 
 La respuesta es:
 
 <font color="red">$ [2,    3,    4,    5,    6,    7,    8,    9,   10,   12,   14,
         15,   16,   18,   20,   21,   22,   24,   25,   26,   27,   28,
         30,   32,   33,   34,   35,   36,   38,   39,   40,   42,   44,
         45,   46,   48,   49,   50,   51,   52,   54,   55,   56,   57,
         58,   60,   62,   63,   64,   65,   66,   68,   69,   70,   72,
         74,   75,   76,   77,   78,   80,   81,   82,   84,   85,   86,
         87,   88,   90,   91,   92,   93,   94,   95,   96,   98,   99,
        100,  102,  104,  105,  106,  108,  110,  111,  112,  114,  115,
        116,  117,  118,  119,  120,  122,  123,  124,  125,  126,  128,
        129,  130,  132,  133,  134,  135,  136,  138,  140,  141,  142,
        144,  145,  146,  147,  148,  150,  152,  153,  154,  155,  156,
        158,  159,  160,  161,  162,  164,  165,  166,  168,  170,  171,
        172,  174,  175,  176,  177,  178,  180,  182,  183,  184,  185,
        186,  188,  189,  190,  192,  194,  195,  196,  198,  200,  201,
        202,  203,  204,  205,  206,  207,  208,  210,  212,  213,  214,
        215,  216,  217,  218,  219,  220,  222,  224,  225,  226,  228,
        230,  231,  232,  234,  235,  236,  237,  238,  240,  242,  243,
        244,  245,  246,  248,  249,  250,  252,  254,  255,  256,  258,
        259,  260,  261,  262,  264,  265,  266,  267,  268,  270,  272,
        273,  274,  275,  276,  278,  279,  280,  282,  284,  285,  286,
        287,  288,  290,  291,  292,  294,  295,  296,  297,  298,  300,
        301,  302,  303,  304,  305,  306,  308,  309,  310,  312,  314,
        315,  316,  318,  320,  321,  322,  324,  325,  326,  327,  328,
        329,  330,  332,  333,  334,  335,  336,  338,  339,  340,  342,
        343,  344,  345,  346,  348,  350,  351,  352,  354,  355,  356,
        357,  358,  360,  362,  363,  364,  365,  366,  368,  369,  370,
        371,  372,  374,  375,  376,  378,  380,  381,  382,  384,  385,
        386,  387,  388,  390,  392,  393,  394,  395,  396,  398,  399,
        400,  402,  404,  405,  406,  408,  410,  411,  412,  413,  414,
        415,  416,  417,  418,  420,  422,  423,  424,  425,  426,  427,
        428,  429,  430,  432,  434,  435,  436,  438,  440,  441,  442,
        444,  445,  446,  447,  448,  450,  452,  453,  454,  455,  456,
        458,  459,  460,  462,  464,  465,  466,  468,  469,  470,  471,
        472,  474,  475,  476,  477,  478,  480,  482,  483,  484,  485,
        486,  488,  489,  490,  492,  494,  495,  496,  497,  498,  500,
        501,  502,  504,  505,  506,  507,  508,  510,  511,  512,  513,
        514,  515,  516,  518,  519,  520,  522,  524,  525,  526,  528,
        530,  531,  532,  534,  535,  536,  537,  538,  539,  540,  542,
        543,  544,  545,  546,  548,  549,  550,  552,  553,  554,  555,
        556,  558,  560,  561,  562,  564,  565,  566,  567,  568,  570,
        572,  573,  574,  575,  576,  578,  579,  580,  581,  582,  584,
        585,  586,  588,  590,  591,  592,  594,  595,  596,  597,  598,
        600,  602,  603,  604,  605,  606,  608,  609,  610,  612,  614,
        615,  616,  618,  620,  621,  622,  623,  624,  625,  626,  627,
        628,  630,  632,  633,  634,  635,  636,  637,  638,  639,  640,
        642,  644,  645,  646,  648,  650,  651,  652,  654,  655,  656,
        657,  658,  660,  662,  663,  664,  665,  666,  668,  669,  670,
        672,  674,  675,  676,  678,  679,  680,  681,  682,  684,  685,
        686,  687,  688,  690,  692,  693,  694,  695,  696,  698,  699,
        700,  702,  704,  705,  706,  707,  708,  710,  711,  712,  714,
        715,  716,  717,  718,  720,  721,  722,  723,  724,  725,  726,
        728,  729,  730,  732,  734,  735,  736,  738,  740,  741,  742,
        744,  745,  746,  747,  748,  749,  750,  752,  753,  754,  755,
        756,  758,  759,  760,  762,  763,  764,  765,  766,  768,  770,
        771,  772,  774,  775,  776,  777,  778,  780,  782,  783,  784,
        785,  786,  788,  789,  790,  791,  792,  794,  795,  796,  798,
        800,  801,  802,  804,  805,  806,  807,  808,  810,  812,  813,
        814,  815,  816,  818,  819,  820,  822,  824,  825,  826,  828,
        830,  831,  832,  833,  834,  835,  836,  837,  838,  840,  842,
        843,  844,  845,  846,  847,  848,  849,  850,  852,  854,  855,
        856,  858,  860,  861,  862,  864,  865,  866,  867,  868,  870,
        872,  873,  874,  875,  876,  878,  879,  880,  882,  884,  885,
        886,  888,  889,  890,  891,  892,  894,  895,  896,  897,  898,
        900,  902,  903,  904,  905,  906,  908,  909,  910,  912,  914,
        915,  916,  917,  918,  920,  921,  922,  924,  925,  926,  927,
        928,  930,  931,  932,  933,  934,  935,  936,  938,  939,  940,
        942,  944,  945,  946,  948,  950,  951,  952,  954,  955,  956,
        957,  958,  959,  960,  962,  963,  964,  965,  966,  968,  969,
        970,  972,  973,  974,  975,  976,  978,  980,  981,  982,  984,
        985,  986,  987,  988,  990,  992,  993,  994,  995,  996,  998,
        999, 1000] $</font>